File: platform_src/android/app/src/main/java/org/elastos/trinity/runtime/TrinityPlugin.java
/*
 * Copyright (c) 2018 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.trinity.runtime;

import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaPreferences;
import org.apache.cordova.CordovaWebView;

import java.io.File;

public class TrinityPlugin extends CordovaPlugin {
    private AppWhitelistPlugin whitelistPlugin = null;
    public String dataPath = null;
    public String appPath = null;
    public String tempPath = null;
    public String configPath = null;
    private AppInfo appInfo = null;
    public AppManager appManager = null;
    protected String appId;

    public void setWhitelistPlugin(AppWhitelistPlugin appWhitelistPlugin) {
        this.whitelistPlugin = appWhitelistPlugin;
    }

    public void setInfo(AppInfo info) {
        this.appInfo = info;
        this.appManager = AppManager.getShareInstance();
        this.appPath = appManager.getAppPath(info);
        this.dataPath = appManager.getDataPath(info.app_id);
        this.configPath = appManager.getConfigPath();
        this.tempPath = appManager.getTempPath(info.app_id);
        this.appId = info.app_id;
    }

    public boolean isAllowAccess(String url) {
        if (whitelistPlugin != null) {
            Boolean ret = whitelistPlugin.shouldAllowNavigation(url);
            if (ret == Boolean.TRUE) {
                return true;
            }
            else {
                return false;
            }
        }
        else {
            return true;
        }
    }

    public boolean isUrlApp() {
        return appInfo.type.equals("url");
    }

    public String getAppPath() {
        return appPath;
    }

    public String getDataPath() {
        return dataPath;
    }

    public String getConfigPath() {
        return configPath;
    }

    public String getTempPath() {
        return tempPath;
    }

    private String getCanonicalDir(String path, String header) throws Exception {
        File file = new File(path);
        path = file.getCanonicalPath();

        if (header.equals(path + "/")) {
            return "";
        }

        if (!header.startsWith("/")) {
            path = path.substring(1);
        }
        if (!path.startsWith(header)) {
            throw new Exception("Dir is invalid!");
        }
        String dir = path.substring(header.length());
        return dir;
    }

    public String getCanonicalPath(String path) throws Exception {
        String ret = null;
        if (path.startsWith("trinity://")) {
            String subPath = path.substring(10);
            String id = null;
            String _appPath = appPath, _dataPath = dataPath, _tempPath = tempPath;
            if (!(path.startsWith("trinity:///"))) {
                int index = subPath.indexOf("/");
                if (index != -1) {
                    id = subPath.substring(0, index);
                    subPath = subPath.substring(index);

                    AppInfo info = appManager.getAppInfo(id);
                    if (info == null) {
                        id = null;
                    }
                    else {
                        _appPath = appManager.getAppPath(info);
                        _dataPath = appManager.getDataPath(info.app_id);
                        _tempPath = appManager.getTempPath(info.app_id);
                    }
                }
            }
            else {
                id = this.appInfo.app_id;
            }

            if (id != null) {
                if (subPath.startsWith("/asset/")) {
                    String dir = getCanonicalDir(subPath, "/asset/");
                    ret = _appPath + dir;
                } else if (subPath.startsWith("/data/")) {
                    String dir = getCanonicalDir(subPath, "/data/");
                    ret = _dataPath + dir;
                } else if (subPath.startsWith("/temp/")) {
                    String dir = getCanonicalDir(subPath, "/temp/");
                    ret = _tempPath + dir;
                }
            }
        }
        else if ((path.indexOf("://") != -1)) {
            if (!(path.startsWith("asset://")) && isAllowAccess(path)) {
                ret = path;
            }
        }
        else if (!path.startsWith("/")) {
            String dir = getCanonicalDir("/asset/" + path, "/asset/");
            ret = appPath + dir;
        }

        if (ret == null) {
            throw new Exception("Dir is invalid!");
        }

        return ret;
    }

    public String getDataCanonicalPath(String path) throws Exception {
        if (!path.startsWith("trinity:///data/"))  {
            throw new Exception("Dir is invalid!");
        }
        return getCanonicalPath(path);
    }

    public String getRelativePath(String path) throws Exception {
        String ret = null;
        if (!(path.startsWith("asset://")) && (path.indexOf("://") != -1)) {
            if (isAllowAccess(path)) {
                ret = path;
            }
        }
        else if (path.startsWith(appPath)) {
            File file = new File(appPath);
            String header = file.getCanonicalPath() + "/";
            ret = "trinity:///asset/" + getCanonicalDir(path, header);
        }
        else if (path.startsWith(dataPath)) {
            File file = new File(dataPath);
            String header = file.getCanonicalPath() + "/";
            ret = "trinity:///data/" + getCanonicalDir(path, header);
        }
        else if (path.startsWith(tempPath)) {
            File file = new File(tempPath);
            String header = file.getCanonicalPath() + "/";
            ret = "trinity:///temp/" + getCanonicalDir(path, header);
        }

        if (ret == null) {
            throw new Exception("Dir is invalid!");
        }
        return ret;
    }

    @Override
    public final void privateInitialize(String serviceName, CordovaInterface cordova,
                                        CordovaWebView webView, CordovaPreferences preferences) {
        if (appInfo == null) {
            setInfo(AppManager.getShareInstance().getLauncherInfo());
        }
        super.privateInitialize(serviceName, cordova, webView, preferences);
    }
}

File: platform_src/ios/Trinity/Classes/TrinityPlugin.m
/*
 * Copyright (c) 2018 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#import "TrinityPlugin.h"
#import "Trinity-Swift.h"
#include <objc/message.h>

@interface TrinityPlugin()
@property (nonatomic, readwrite, strong) WhitelistFilter* filter;
@property (nonatomic, readwrite) BOOL checkAuthority;
@property (nonatomic, readwrite, copy) NSString* pluginName;
@property (nonatomic, readwrite) NSString* appPath;
@property (nonatomic, readwrite) NSString* dataPath;
@property (nonatomic, readwrite) NSString* configPath;
@property (nonatomic, readwrite) NSString* tempPath;
@end

@implementation TrinityPlugin

@synthesize filter;
@synthesize checkAuthority;
@synthesize pluginName;
@synthesize appPath;
@synthesize dataPath;
@synthesize configPath;
@synthesize tempPath;

- (void)setInfo:(CDVPlugin *)filter checkAuthority:(BOOL)check
        appPath:(NSString*)appPath dataPath:(NSString*)dataPath
     configPath:(NSString*)configPath tempPath:(NSString*)tempPath {
    self.filter = (WhitelistFilter*)filter;
    self.checkAuthority = check;
    self.appPath = appPath;
    self.dataPath = dataPath;
    self.configPath = configPath;
    self.tempPath = tempPath;
}

- (BOOL)isAllowAccess:(NSString *)url {
    return [self.filter shouldAllowNavigation:url];
}

- (NSString*)getAppPath {
    return appPath;
}

- (NSString*)getDataPath {
    return dataPath;
}

- (NSString*)getTempPath {
    return tempPath;
}

- (NSString*)getConfigPath {
    return configPath;
}

- (void)setError:(NSError * _Nullable *)error {
    if (error == NULL) {
        return;
    }
    NSString *domain = @"";
    NSString *desc = NSLocalizedString(@"Dir is invalid!", @"");
    NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : desc };

    *error = [NSError errorWithDomain:domain
                                 code:-101
                             userInfo:userInfo];
}

- (NSString*)getCanonicalDir:(NSString *)path header:(NSString*)header error:(NSError * _Nullable *)error {
    path = [path stringByStandardizingPath];

    if ([header isEqualToString: [path stringByAppendingString:@"/"]]) {
        return @"";
    }

    if (![header hasPrefix:@"/"]) {
        path = [path substringFromIndex:1];
    }
    if (![path hasPrefix:header]) {
        return NULL;
    }
    NSString* dir = [path substringFromIndex:header.length];
    return dir;
}

- (NSString*)getCanonicalPath:(NSString*)path error:(NSError * _Nullable *)error {
    NSString* ret = NULL;
    if (path == NULL) {
        [self setError:error];
        return NULL;
    }

    if ([path hasPrefix:@"trinity:///asset/"]) {
        NSString* dir = [self getCanonicalDir:[path substringFromIndex:10] header:@"/asset/" error:error];
        if (dir != NULL) {
            ret = [appPath stringByAppendingString:dir];
        }
    }
    else if ([path hasPrefix:@"trinity:///data/"]) {
        NSString* dir = [self getCanonicalDir:[path substringFromIndex:10] header:@"/data/" error:error];
        if (dir != NULL) {
            ret = [dataPath stringByAppendingString:dir];
        }
    }
    else if ([path hasPrefix:@"trinity:///temp/"]) {
        NSString* dir = [self getCanonicalDir:[path substringFromIndex:10] header:@"/temp/" error:error];
        if (dir != NULL) {
            ret = [tempPath stringByAppendingString:dir];
        }
    }
    else if ([path rangeOfString:@"://"].length > 0) {
        if (![path hasPrefix:@"asset://"] && [self.filter shouldAllowNavigation:path]) {
            ret = path;
        }
    }
    else if (![path  hasPrefix:@"/"]) {
        NSString* dir = [self getCanonicalDir:[@"/asset/" stringByAppendingString:path] header:@"/asset/" error:error];
        if (dir != NULL) {
            ret = [appPath stringByAppendingString:dir];
        }
    }

    if (ret == NULL) {
        [self setError:error];
        return NULL;
    }

    return ret;
}

- (NSString*)getDataCanonicalPath:(NSString*)path error:(NSError * _Nullable *)error {
    if (path ==NULL || ![path hasPrefix:@"trinity:///data/"]) {
        [self setError:error];
        return nil;
    }
    return [self getCanonicalPath:path error:error];
}

- (NSString*)getRelativePath:(NSString*)path error:(NSError * _Nullable *)error {
    NSString* ret = nil;
    if (path == NULL) {
        [self setError:error];
        return nil;
    }

    if (![path hasPrefix:@"asset://"] && [path rangeOfString:@"://"].length > 0) {
        if ([self.filter shouldAllowNavigation:path]) {
            ret = path;
        }
    }
    else if ([path hasPrefix:appPath]) {
        NSString* dir = [self getCanonicalDir:path  header:appPath error:error];
        if (dir != nil) {
            ret = [@"trinity:///asset/" stringByAppendingString:dir];
        }
    }
    else if ([path hasPrefix:dataPath]) {
        NSString* dir = [self getCanonicalDir:path  header:dataPath error:error];
        if (dir != nil) {
            ret = [@"trinity:///data/" stringByAppendingString:dir];
        }
    }
    else if ([path hasPrefix:tempPath]) {
        NSString* dir = [self getCanonicalDir:path  header:tempPath error:error];
        if (dir != nil) {
            ret = [@"trinity:///temp/" stringByAppendingString:dir];
        }
    }
    if (ret == nil) {
        [self setError:error];
        return nil;
    }

    return ret;
}

//- (BOOL)execute:(CDVInvokedUrlCommand*)command
//{
//    NSString* methodName = [NSString stringWithFormat:@"%@:", command.methodName];
//    SEL normalSelector = NSSelectorFromString(methodName);
//    if ([self respondsToSelector:normalSelector]) {
//        ((void (*)(id, SEL, id))objc_msgSend)(self, normalSelector, command);
//        return YES;
//    } else {
//        NSString* msg = [NSString stringWithFormat:@"ERROR: Method '%@' not defined in Plugin '%@'", methodName, command.className];
//        NSLog(@"%@", msg);
//        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:msg];
//        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
//        return YES;
//    }
//}
//
//- (BOOL)trinityExecute:(CDVInvokedUrlCommand*)command
//{
//    if (checkAuthority) {
//        int authority = [self.filter getPluginAuthority:self.pluginName
//                                        trinityPlugin: self
//                                        invokedUrlCommand: command];
//        if (authority == AppInfo.AUTHORITY_ASK) {
//            NSString* msg = [NSString stringWithFormat:@"Plugin:'%@' have not run authority.", pluginName];
//            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:msg];
//            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
//            return YES;
//        }
//        else if (authority == AppInfo.AUTHORITY_NOINIT || authority == AppInfo.AUTHORITY_ASK) {
//            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_NO_RESULT];
//            [result setKeepCallbackAsBool:YES];
//            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
//            return YES;
//        }
//    }
//
//    return [self execute:command];
//}
@end

